### PR TITLE
support `reg_byte` registers

### PR DESCRIPTION
**Pull request by folkertdev**
fixes https://github.com/rust-lang/rustc_codegen_gcc/issues/485

I believe (see the issue) that the `b` suffix can just be trimmed, without causing invalid behavior.

I also refactored a bit to reduce the level of nesting, given that we probably need much more logic here to support more register classes that don't translate 1-to-1 between llvm and gcc. Hopefully the rustc assembly tests will guide us towards the edge cases.

### CODE REVIEW

File: src/asm.rs [hunk 1]
```diff
@@ -611,114 +611,126 @@ fn estimate_template_length(
 }
 
 /// Converts a register class to a GCC constraint code.
-fn reg_to_gcc(reg: InlineAsmRegOrRegClass) -> ConstraintOrRegister {
-    let constraint = match reg {
-        // For vector registers LLVM wants the register name to match the type size.
```
Comment (folkertdev) [inline src/asm.rs:616]: this comment is lost right now, I'm not sure what it was supposed to say, and why what llvm does is relevant?

Comment (antoyo) [reply src/asm.rs:616]: Oh, that was copy-pasted from the LLVM codegen, but I forgot to change that to GCC.

Comment (folkertdev) [reply src/asm.rs:616]: ok, but what does this comment mean? What type is talked about here? it just doesn't seem like the right place for this comment?

Comment (antoyo) [reply src/asm.rs:616]: It comes from [here](https://github.com/rust-lang/rust/blob/6ac88786155774da429213a1362d4cb085047af3/compiler/rustc_codegen_llvm/src/asm.rs#L587) originally.
Maybe the type is the `layout` parameter that we don't have in cg_gcc.

Comment (folkertdev) [reply src/asm.rs:616]: yeah I think that's right, e.g. 

```rust
  match layout.size.bytes() {
                        64 => 'z',
                        32 => 'y',
                        _ => 'x',
```

this really does pick the register width based on a type. Do you think GCC will have to do the same? So far that does not seem necessary.

Comment (antoyo) [reply src/asm.rs:616]: I do not know. I would need to check if needed.

File: src/asm.rs [hunk 2]
```diff
@@ -611,114 +611,126 @@ fn estimate_template_length(
...
-                        name => name,
-                    });
+fn explicit_reg_to_gcc(reg: InlineAsmReg) -> &'static str {
+    // For explicit registers, we have to create a register variable: https://stackoverflow.com/a/31774784/389119
+    match reg {
+        InlineAsmReg::X86(reg) => {
+            // TODO(antoyo): add support for vector register.
+            match reg.reg_class() {
+                X86InlineAsmRegClass::reg_byte => {
+                    // see https://github.com/rust-lang/rustc_codegen_gcc/issues/485
```
Comment (antoyo) [inline src/asm.rs:632]: Could you also explain in the comment why we do this?

File: tests/run/asm.rs [hunk 3]
```diff
@@ -205,6 +205,29 @@ fn asm() {
...
+    unsafe {
+        core::arch::asm!(
+            "",
+            out("al") _,
+            out("bl") _,
+            out("cl") _,
+            out("dl") _,
+            out("sil") _,
+            out("dil") _,
+            // out("bpl") _, // frame pointer
```
Comment (antoyo) [inline tests/run/asm.rs:220]: Can this line be removed?

**Review by antoyo — APPROVED** [review]

**Comment by antoyo** [comment]
Is this ready to be merged?

**Comment by folkertdev** [comment]
yes

**Comment by antoyo** [comment]
Thanks for your contribution!
